oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java

```diff
@@ -748,6 +931,43 @@ String createAssertionForIdToken(
     }
   }
 
+  /**
+   * Self signed JWT uses uri as audience, and it should have the "https://{host}/" format. For
+   * instance, if the uri is "https://compute.googleapis.com/compute/v1/projects/", then this
+   * function returns "https://compute.googleapis.com/".
+   */
+  @VisibleForTesting
+  static URI getUriForSelfSignedJWT(URI uri) {
+    if (uri == null) {
+      return uri;
+    }
+    try {
+      return new URI(uri.getScheme(), uri.getHost(), "/", null);
```

Comment:
There's no possibility of a port or a user info here? Or if there is, you want to strip that?

Reply:
Yep, they should be stripped if there are any. We are constructing audience from uri, and it should look like `http(s)://compute.googleapis.com/`